test-app/app/build.gradle

```diff
@@ -1,127 +1,452 @@
-def runOnDevice = project.hasProperty("runOnDevice");
-def runOnDeviceOrEmulator = runOnDevice ? "-d" : "-e";
+////////////////////////////////////////////////////////////////////////////////////
+///////////////////////////// CONFIGURATION PHASE //////////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////
 
-task deleteDist(type: Delete) {
-    doFirst {
-        delete "$rootDir/dist"
+task addNativeScriptRuntimePackageDependency {
+    def useV8Symbols = nativescriptDependencies.any {
+        def packageJsonPath = file("$rootDir/${it.directory}/$PACKAGE_JSON")
+        def packageJson = new JsonSlurper().parseText(packageJsonPath.text)
+        return packageJson.nativescript.useV8Symbols
+    }
+
+    if (!externalRuntimeExists) {
+        def runtime = useV8Symbols ? "nativescript-regular" : "nativescript-optimized"
+        println "\t + adding nativescript runtime package dependency: $runtime"
+        project.dependencies.add("compile", [name: runtime, ext: "aar"])
+    } else {
+        project.dependencies.add("compile", project(':runtime'))
     }
 }
 
-task installApk(type: Exec) {
-    doFirst {
-        println "Attempting to install buit apk"
+task addDependenciesFromNativeScriptPlugins {
+    nativescriptDependencies.each { dep ->
+        def aarFiles = fileTree(dir: file("$rootDir/${dep.directory}/$PLATFORMS_ANDROID"), include: ["**/*.aar"])
+        aarFiles.each { aarFile ->
+            def length = aarFile.name.length() - 4
+            def fileName = aarFile.name[0..<length]
+            println "\t + adding aar plugin dependency: " + aarFile.getAbsolutePath()
+            project.dependencies.add("compile", [name: fileName, ext: "aar"])
+        }
 
-        if (isWinOs) {
-            commandLine "cmd", "/c", "node", "$rootDir\\tools\\deploy-apk.js", "$rootDir\\app\\build\\outputs\\apk\\app-debug.apk", runOnDeviceOrEmulator
-        } else {
-            commandLine "node", "$rootDir/tools/deploy-apk.js", "$rootDir/app/build/outputs/apk/app-debug.apk", runOnDeviceOrEmulator
+        def jarFiles = fileTree(dir: file("${dep.directory}/$PLATFORMS_ANDROID"), include: ["**/*.jar"])
+        jarFiles.each { jarFile ->
+            println "\t + adding jar plugin dependency: " + jarFile.getAbsolutePath()
         }
+
+        project.dependencies.add("compile", jarFiles)
     }
+}
 
+tasks.whenTaskAdded({ org.gradle.api.DefaultTask currentTask ->
+    if (currentTask =~ /generate.+BuildConfig/) {
+        currentTask.finalizedBy(extractAllJars)
+        extractAllJars.finalizedBy(collectAllJars)
+    }
+    if (currentTask =~ /compile.+JavaWithJavac/) {
+        currentTask.dependsOn(runSbg)
+        currentTask.finalizedBy(ensureMetadataOutDir)
+        ensureMetadataOutDir.finalizedBy(buildMetadata)
+    }
+    if (currentTask.equals("assembleDebug") || currentTask.equals("assembleRelease")) {
```

Comment:
Should we do `.+BuildConfig` matching similar to the one above to ensure we don't miss a case, like flavor-specific assemble tasks?

Reply:
good point, will do